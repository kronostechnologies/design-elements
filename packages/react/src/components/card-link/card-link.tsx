import { VoidFunctionComponent } from 'react';
import { Link } from 'react-router-dom';
import styled from 'styled-components';
import { focus } from '../../utils/css-state';
import { Icon } from '../icon/icon';

const StyledIcon = styled(Icon)`
    color: currentColor;
    height: var(--size-1x);
    position: absolute;
    right: var(--spacing-3x);
    top: 50%;
    transform: translateY(-50%);
    width: var(--size-1x);
`;

const StyledLink = styled(Link)`
    background-color: ${({ theme }) => theme.component['card-link-background-color']};
    border: 1px solid ${({ theme }) => theme.component['card-link-border-color']};
    border-radius: var(--border-radius-2x);
    box-sizing: border-box;
    color: ${({ theme }) => theme.component['card-link-text-color']};
    display: block;
    font-size: 1rem;
    font-weight: var(--font-semi-bold);
    line-height: 1.5rem;
    overflow: hidden;
    padding: var(--spacing-3x) var(--spacing-5x) var(--spacing-3x) var(--spacing-3x);
    position: relative;
    text-decoration: none;
    text-overflow: ellipsis;
    white-space: nowrap;
    width: 100%;

    &:hover {
        background-color: ${({ theme }) => theme.component['card-link-hover-background-color']};
        border-color: ${({ theme }) => theme.component['card-link-hover-box-shadow-inset-color']};
    }

    ${focus};
`;

interface CardLinkProps {
    className?: string;
    href: string;
    label: string;
    /** When true, clicking the link will replace the current entry in the history stack instead of adding a new one. */
    replace?: boolean;
}

export const CardLink: VoidFunctionComponent<CardLinkProps> = ({
    className, href, label, replace,
}) => (
    <StyledLink className={className} replace={replace} title={label} to={href}>
        {label}
        <StyledIcon aria-hidden name="chevronRight" size="16" />
    </StyledLink>
);
