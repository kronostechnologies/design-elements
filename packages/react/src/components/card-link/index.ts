export { CardLink } from './card-link';
export type { CardLinkProps } from './card-link';
