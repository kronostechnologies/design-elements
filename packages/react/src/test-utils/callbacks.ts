export function doNothing(): void {
    // do nothing
}
