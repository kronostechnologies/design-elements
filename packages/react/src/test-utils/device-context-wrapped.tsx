import React, { ReactElement } from 'react';
import { DeviceContextProvider, DeviceType } from '../components/device-context-provider/device-context-provider';

export function DeviceContextWrapped(children: ReactElement, device?: DeviceType): ReactElement {
    return (
        <DeviceContextProvider staticDevice={device}>
            {children}
        </DeviceContextProvider>
    );
}
