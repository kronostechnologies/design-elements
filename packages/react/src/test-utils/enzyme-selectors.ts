import { ReactWrapper, ShallowWrapper } from 'enzyme';

enum ModifierEnum {
    StartsWith = '^',
    EndsWith = '$',
    ContainsOptionalHyphen = '|',
    Contains = '*',
    ContainsWord = '~',
    None = '',
}
type SelectorModifier = `${ModifierEnum}`;

function filterHtmlNodesOnly<W extends ReactWrapper | ShallowWrapper>(
    nodes: W,
): W {
    return nodes
        .hostNodes()
        .filterWhere((x) => x.name() !== null && !x.name().startsWith('styled.')) as W;
}

export function findByTestId<W extends ReactWrapper | ShallowWrapper>(
    wrapper: W,
    testId: string,
    options?: {
        modifier?: SelectorModifier,
        nodeName?: string,
        // Don't use with shallow wrapper as it contains styled node over the html nodes.
        htmlNodesOnly?: boolean,
    },
): W {
    const modifier = options?.modifier ?? '';

    let nodes = wrapper.find(`[data-testid${modifier}="${testId}"]`) as W;

    if (options?.nodeName) {
        nodes = nodes.filterWhere((x) => x.name() === options.nodeName) as W;
    }
    if (options?.htmlNodesOnly) {
        nodes = filterHtmlNodesOnly(nodes);
    }

    return nodes;
}

export function getByTestId<W extends ReactWrapper | ShallowWrapper>(
    wrapper: W,
    testId: string,
    options?: {
        modifier?: SelectorModifier,
        nodeName?: string,
        // Don't use with shallow wrapper as it contains styled node instead of the html nodes.
        htmlNodesOnly?: boolean,
    },
): W {
    return findByTestId(wrapper, testId, options).at(0) as W;
}
