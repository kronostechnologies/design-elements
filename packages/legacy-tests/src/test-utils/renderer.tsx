import { CommonWrapper, mount, MountRendererProps, ReactWrapper, render } from 'enzyme';
import { act, Component, FunctionComponent, PropsWithChildren, ReactElement } from 'react';
import { MemoryRouter } from 'react-router-dom';
import { DesignSystem, DesignSystemProps } from '~/components/design-system';
import { ThemeWrapper } from '~/components/theme-wrapper/theme-wrapper';

export const AllProviders: FunctionComponent<PropsWithChildren<DesignSystemProps>> = ({
    children,
    language,
    theme,
    staticDevice,
}) => (
    <MemoryRouter future={{ v7_startTransition: false, v7_relativeSplatPath: false }}>
        <DesignSystem language={language} staticDevice={staticDevice} theme={theme}>
            {children}
        </DesignSystem>
    </MemoryRouter>
);

interface WrappingComponentProps {
    wrappingComponentProps?: DesignSystemProps;
}

type Options = MountRendererProps & WrappingComponentProps;

export function mountWithProviders<C extends Component, P = C['props'], S = C['state']>(
    component: ReactElement<P>,
    options: Options = {},
): ReactWrapper<P, S, C> {
    return mount(component, {
        ...options,
        wrappingComponent: AllProviders as Options['wrappingComponent'],
    });
}

export function mountWithTheme<C extends Component, P = C['props'], S = C['state']>(
    component: ReactElement<P>,
    options?: MountRendererProps,
): ReactWrapper<P, S, C> {
    return mount(component, {
        wrappingComponent: ThemeWrapper as MountRendererProps['wrappingComponent'],
        ...options,
    });
}

export function renderWithTheme(
    component: ReactElement,
): cheerio.Cheerio {
    return render(
        <ThemeWrapper>
            {component}
        </ThemeWrapper>,
    );
}

export async function actUpdate<C extends Component, P = C['props'], S = C['state']>(
    wrapper: CommonWrapper<P, S, C>,
): Promise<void> {
    await act(async () => {
        wrapper.update();
    });
}

export async function actAndWaitForEffects<C extends Component, P = C['props'], S = C['state']>(
    wrapper: ReactWrapper<P, S, C>,
    action: () => void,
): Promise<ReactWrapper<P, S, C>> {
    await act(async () => action());
    await actUpdate(wrapper);
    return wrapper;
}
