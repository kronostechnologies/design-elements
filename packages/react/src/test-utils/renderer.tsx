import { render as testingLibRender, RenderResult } from '@testing-library/react';
import { FunctionComponent, PropsWithChildren, ReactElement } from 'react';
import { MemoryRouter } from 'react-router-dom';
import { DesignSystem, DesignSystemProps } from '../components/design-system';
import { DeviceType } from '../components/device-context-provider/device-context-provider';

export const AllProviders: FunctionComponent<PropsWithChildren<DesignSystemProps>> = ({
    children,
    language,
    theme,
    staticDevice,
}) => (
    <MemoryRouter future={{ v7_startTransition: false, v7_relativeSplatPath: false }}>
        <DesignSystem language={language} staticDevice={staticDevice} theme={theme}>
            {children}
        </DesignSystem>
    </MemoryRouter>
);

export function renderWithProviders(
    component: ReactElement,
    device?: DeviceType,
): RenderResult {
    return testingLibRender(<AllProviders staticDevice={device}>{component}</AllProviders>);
}

export function rerenderPortalWithProviders(
    component: ReactElement,
    rerender: RenderResult['rerender'],
    device?: DeviceType,
): void {
    rerender(<AllProviders staticDevice={device}>{component}</AllProviders>);
}
