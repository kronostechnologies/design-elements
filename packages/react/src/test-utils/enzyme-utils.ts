import { ReactWrapper } from 'enzyme';
import { act } from 'react-dom/test-utils';

export async function waitForComponentToPaint<P = {}>(
    wrapper: ReactWrapper<P>,
    amount = 0,
): Promise<void> {
    await act(async () => {
        await new Promise((resolve) => setTimeout(resolve, amount));
        wrapper.update();
    });
}

export function expectFocusToBeOn(element: ReactWrapper): void {
    expect(document.activeElement).toBe(element.getDOMNode());
}
