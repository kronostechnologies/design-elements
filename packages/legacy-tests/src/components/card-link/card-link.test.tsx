import { BrowserRouter as Router } from 'react-router-dom';
import { CardLink } from '~/components/card-link/card-link';
import { renderWithTheme } from '../../test-utils/renderer';

describe('CardLink', () => {
    it('Matches Snapshot', () => {
        const tree = renderWithTheme(
            <Router>
                <CardLink label="Link Label" href="/" />
            </Router>,
        );

        expect(tree).toMatchSnapshot();
    });
});
