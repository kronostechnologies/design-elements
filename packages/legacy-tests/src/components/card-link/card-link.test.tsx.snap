// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`CardLink Matches Snapshot 1`] = `
.c1 {
  color: currentColor;
  height: var(--size-1x);
  position: absolute;
  right: var(--spacing-3x);
  top: 50%;
  -webkit-transform: translateY(-50%);
  -ms-transform: translateY(-50%);
  transform: translateY(-50%);
  width: var(--size-1x);
}

.c0 {
  background-color: #FFFFFF;
  border: 1px solid #878F9A;
  border-radius: var(--border-radius-2x);
  box-shadow: 0 1px 4px 0 rgb(0 0 0 / 0.2);
  box-sizing: border-box;
  color: #1B1C1E;
  display: block;
  font-size: 1rem;
  font-weight: var(--font-semi-bold);
  line-height: 1.5rem;
  overflow: hidden;
  padding: var(--spacing-3x) var(--spacing-5x) var(--spacing-3x) var(--spacing-3x);
  position: relative;
  -webkit-text-decoration: none;
  text-decoration: none;
  text-overflow: ellipsis;
  white-space: nowrap;
  width: 100%;
}

.c0:hover {
  background-color: #DBDEE1;
  border-color: #000000;
}

.c0 {
  outline: 2px solid transparent;
  outline-offset: -2px;
}

.c0:focus-visible {
  box-shadow: 0 0 0 2px #006296;
  outline: 2px solid #84C6EA;
  outline-offset: -2px;
}

<a
  class="c0"
  href="/"
  title="Link Label"
>
  Link Label
  <svg
    aria-hidden="true"
    class="c1"
    color="currentColor"
    focusable="false"
    height="16"
    width="16"
  />
</a>
`;
